### ISSUE TITLE
Validation of periodic builds setup on Nordix Infrastructure for selected ETOS repos

### ISSUE THREAD

**Issue by fdegir** [eiffel-community/etos#54]
### Description

Periodic builds are currently being setup for selected ETOS projects/repositories on Nordix Infrastructure as part of Eiffel Infrastructure and CI/CD establishment work [Phase 2](https://hackmd.io/X3Nrxe4MRdSJI_lsTkSwiQ?view#Phase-2-Build-Rollout). List of projects is

- etos-api
- etos-environment-provider
- etos-suite-runner
- etos-suite-starter

It is important the build process is reviewed by the project to ensure it is correct.

The temporary jobs for these projects are available on [Nordix Jenkins](https://jenkins.nordix.org/view/Eiffel/). (Jobs named tmp-etos-*)

The builds for etos-test-runner-containers will be worked on once the build process of above projects/repos are reviewed.

Here are more details.

- The job created for these projects on Nordix Jenkins run against the master branches of the git repositories and attempts building container images using podman
- The jobs regularly poll the repo for commits and get triggered if one or more commits are found. Please note that this can be adjusted depending on the project needs and triggers can be updated such as to trigger based on new PRs.
- Upon completion of the builds, the container images will be pushed to eiffel project on Nordix Container Image Registry (Please reach out to @fdegir if you need account on the registry.)

Build process.

- When the jobs get triggered, project repos are cloned and the last commit on the tip of master branch is checked out
- Once the clone is ready, jobs proceed with podman build using the command below
`podman build --log-level error --tag registry.nordix.org/eiffel/<etos-repo-name>:latest .`
- Image tag is set to latest since this is a build against the tip of master branch
- The complete build script is available on Nordix CICD Repository.
- Upon completion of the build, the container image is pushed to eiffel project on Nordix Container Image Registry
- The image is available for anonymous users and can be pulled using either one of the commands below
```
podman pull registry.nordix.org/eiffel/<etos-repo-name>:latest
docker pull registry.nordix.org/eiffel/<etos-repo-name>:latest
```

What we need from you is

- Look at the build logs of the temporary jobs and let us know if you notice anything that looks strange
- Review the change on [Nordix Gerrit](https://gerrit.nordix.org/c/infra/cicd/+/6266) that creates the real jobs. (you need to login to Nordix Gerrit using your Google account in order to be able to review the code)
- Correct the build process or the command used for the project build if they are incorrect
- If possible, pull the image and verify it works
```
podman pull registry.nordix.org/eiffel/<etos-repo-name>:latest
docker pull registry.nordix.org/eiffel/<etos-repo-name>:latest
```
- Provide your opinion about what type of builds your project needs - PR validation before merge, builds after merge, builds after release, periodic builds and anything else you may think if.

Please note that upcoming phases will enable testing as well and we will reach out to you when the time comes.

### Motivation

Setup builds for the project

### Exemplification

N/A

### Benefits

Project will have automated builds available

### Possible Drawbacks

N/A

**Comment by t-persson** [comment]
Thank you.

I cannot see the tmp-etos-* jobs in Jenkins.

At the moment we want to be responsible for our own release work when it comes to production ETOS so nothing should be done on releases.

When it comes to the jobs I would like to see docker built after merge and published with the :dev tag as this is how we work today. This could be changed to "latest" if we want to @fredjn  @andjoe-axis  has to chip in on that.

As I wrote in the gerrit review: We added another dockerfile in the etos-environment-provider today.

**Comment by fdegir** [comment]
> I cannot see the tmp-etos-* jobs in Jenkins.

Sorry, my mistake. 
The temporary jobs were created manually in a wrong way and they were deleted by the job updating jobs.
They are there now: https://jenkins.nordix.org/view/Eiffel/

> At the moment we want to be responsible for our own release work when it comes to production ETOS so nothing should be done on releases.

Got it.]

But still, would you want to have new set of jobs, _<etos-project>-release-ubuntu2004-master_, which gets triggered when you do a release on GitHub so the jobs could build the new image, apply the tag to them based on what you used in GitHub during the release, and push them to Nordix Harbor eiffel project?

Or, do you plan to build/tag/push those release images manually?

> When it comes to the jobs I would like to see docker built after merge and published with the :dev tag as this is how we work today. This could be changed to "latest" if we want to @fredjn @andjoe-axis has to chip in on that.

The tag _:dev_ works as well. Another thing that comes to my mind is that if we could somehow align the tags used for images that are built after merges for all the projects, this could help us when we want to bring up a Kubernetes cluster and test those versions. Unfortunately I haven't received any feedback from the maintainers of other projects with regards to the builds we created.

> As I wrote in the gerrit review: We added another dockerfile in the etos-environment-provider today.

Thanks for the review. I'll fix this in the change on Gerrit.

**Comment by fdegir** [comment]
An update; the jobs for subset of ETOS repos are created https://jenkins.nordix.org/view/Eiffel/ and the images built/tagged/pushed to eiffel project on Nordix Harbor

The next change creates jobs for etos-test-runner-containers.

Once all these jobs are in place, the intention is to update their triggers to run after merge which I'll need you help to configure repos and enable webhook and give permission to Nordix Infra user.

**Comment by t-persson** [comment]
For now, we will create everything for our production releases ourselves (including the building and pushing of images).
We probably want to change this in the future and move over step by step to nordix :)

The ":dev" tag is better than ":latest" because it's clearer that it's not considered a stable image. ":latest" should be used for the latest production/stable release.

We will start looking at the jobs in jenkins to make sure it works. Thank you so much for doing this work!

**Comment by t-persson** [comment]
@fdegir 
The test-runner-containers build requires build-args input for building. Is this possible? If so, can we add these ourselves? Can we add them as an input parameter with default values?

**Comment by fdegir** [comment]
@t-persson yes, it is possible. The jobs currently use the common job template and scripts.

We can update the common job template and set versions of all tools as parameters which become available to build scripts as environment variables assuming setting these parameters do not cause issues for other ETOS builds.

If so, could you please pass the values for below variables so we can update the jobs?
- ETR_PYTHON_VERSION
- GO_VERSION
- PYTHON_VERSION
- RUST_VERSION

**Comment by t-persson** [comment]
We pass the values as --build-args to the docker build process.
As I said in the other issue over at the community repository I think that we should not automate those builds. Better for us to run them manually with input parameters ourselves.

DEBIAN_RELEASE=buster
GO_VERSION=1.15.3 (latest version on golang.org)
PYTHON_VERSION=3.8.3 (not latest, but the one we've verified)
RUST_VERSION=1.47.0 (latest on https://www.rust-lang.org)
ETR_PYTHON_VERSION=3.8.3 (not latest, but the one we've verified)
